src/hevm/src/EVM/Keccak.hs

```diff
@@ -0,0 +1,81 @@
+{-# LANGUAGE DataKinds #-}
+{- |
+    Module: EVM.Keccak
+    Description: Expr passes to determine Keccak assumptions
+-}
+
+module EVM.Keccak (keccakInj, keccakMin) where
+
+import Prelude hiding (Word, LT, GT)
+
+import Data.Set (Set)
+import qualified Data.Set as Set
+import Control.Monad.State
+
+import EVM.Types
+import EVM.Traversals
+
+
+data BuilderState = BuilderState
+  { keccaks :: Set (Expr EWord) }
+  deriving (Show)
+
+initState :: BuilderState
+initState = BuilderState { keccaks = Set.empty }
+
+go :: forall a. Expr a -> State BuilderState (Expr a)
+go = \case
+  e@(Keccak b) -> do
+    s <- get
+    put $ s{keccaks=Set.insert e (keccaks s)}
+    pure e
+  e -> pure e
+
+findKeccakExpr :: forall a. Expr a -> State BuilderState (Expr a)
+findKeccakExpr e = mapExprM go e
+
+findKeccakProp :: Prop -> State BuilderState Prop
+findKeccakProp p = mapPropM go p
+
+findKeccakPropsExprs :: forall a. [Prop] -> [Expr Buf]  -> [Expr Storage]-> State BuilderState ()
+findKeccakPropsExprs ps bufs stores = do
+  mapM findKeccakProp ps;
+  mapM findKeccakExpr bufs;
+  mapM findKeccakExpr stores;
+  pure ()
+
+
+combine :: [a] -> [(a,a)]
+combine lst = combine' lst []
+  where
+    combine' [] acc = concat acc
+    combine' (x:xs) acc =
+      let xcomb = [ (x, y) | y <- xs] in
+      combine' xs (xcomb:acc)
+
+-- | Takes a list of Props, finds all Keccak occurences and generates
+-- assumptions that all keccak invocations never produce an output less than
+-- 50. This is a probabalistic assumption, but it's one that the solidity
+-- compiler relies on (to avoid mapping / dynamic array collisions with
+-- "normal" storage slots), and presenting these results as counterexamples is
+-- not helpful to end users.
+keccakMin :: [Prop] -> [Expr Buf] -> [Expr Storage] -> [Prop]
+keccakMin ps bufs stores =
+  let (_, st) = runState (findKeccakPropsExprs ps bufs stores) initState in
+  fmap minProp (Set.toList (keccaks st))
+  where
+    minProp :: Expr EWord -> Prop
+    minProp k@(Keccak _) = PGT k (Lit 50)
+    minProp _ = error "Internal error: expected keccak expression"
+
+-- | Takes a list of Props, finds all Keccak occurences and generates
+-- Keccak injectivity assumptions for all unique pairs of Keccak calls
+keccakInj :: forall a. [Prop] -> [Expr Buf]  -> [Expr Storage] -> [Prop]
```

Comment:
Maybe we can have a function that generates both keccak injectivity and min assertions to avoid computing the set of keccak calls twice

Reply:
done